### ISSUE TITLE
Tensor subclasses cannot interpose make_dual via __torch_dispatch__

### ISSUE THREAD

**Issue by ezyang** [pytorch/pytorch#73539]
### 🐛 Describe the bug

Issue has been modified, scroll down

----

```import torch
from torch.utils._pytree import tree_map
import torch.autograd.forward_ad as fwAD

class WrapperTensor(torch.Tensor):
    @staticmethod
    def __new__(cls, e):
        r = torch.Tensor._make_wrapper_subclass(cls, e.shape, dtype=e.dtype, requires_grad=False)
        r.elem = e
        return r

    __torch_function__ = torch._C._disabled_torch_function_impl

    def __str__(self):
        return f'WrapperTensor({self.elem})'

    @classmethod
    def __torch_dispatch__(cls, func, types, args=(), kwargs=None):  # type: ignore
        def unwrap(e):
            if isinstance(e, WrapperTensor):
                return e.elem
            else:
                return e

        def wrap(e):
            if isinstance(e, torch.Tensor):
                return WrapperTensor(e)
            else:
                return e

        return tree_map(wrap, func(*tree_map(unwrap, args), **tree_map(unwrap, kwargs)))

primal = torch.ones(1)
tangent = torch.ones(1)

with fwAD.dual_level():
    y = fwAD.make_dual(WrapperTensor(primal), tangent)
    print(y)
```

This prints

```
WrapperTensor(tensor([1.]))
```

### Versions

master

cc @ezyang @albanD @zou3519 @gqchen @pearu @nikitaved @soulitzer @Lezcano @Varal7

**Comment by soulitzer** [comment]
I think this is just a printing issue, running `fwAD.unpack_dual(y).tangent is tangent` returns True

**Comment by ezyang** [comment]
Yeah, I agree with this assessment. Very odd!

**Comment by albanD** [comment]
Not sure what the problem is here?
You override the `__str__` for your subclass to be `f'WrapperTensor({self.elem})'`. So it indeed only prints `'WrapperTensor'` followed by the plain Tensor that you wrap.

This links to Joel's discussion about printing though: how can we provide a good default printing so that users do not hide this information mistakenly. (@jbschlosser did you open an issue about this?)

**Comment by ezyang** [comment]
So it turns out that there is a deeper problem, which is what lead me to not realize this was a printing problem. The following modified program makes matters clearer:

```
import torch
from torch.utils._pytree import tree_map
import torch.autograd.forward_ad as fwAD

from utils import no_dispatch

class WrapperTensor(torch.Tensor):
    @staticmethod
    def __new__(cls, e):
        r = torch.Tensor._make_subclass(cls, e)
        r.elem = e
        return r

    __torch_function__ = torch._C._disabled_torch_function_impl

    def __repr__(self):
        with no_dispatch():
            return f'WrapperTensor(self={super().__repr__()}, elem={self.elem})'

    @classmethod
    def __torch_dispatch__(cls, func, types, args=(), kwargs=None):  # type: ignore
        print(func)
        def unwrap(e):
            if isinstance(e, WrapperTensor):
                return e.elem
            else:
                return e

        def wrap(e):
            if isinstance(e, torch.Tensor):
                return WrapperTensor(e)
            else:
                return e

        return tree_map(wrap, func(*tree_map(unwrap, args), **tree_map(unwrap, kwargs)))

primal = torch.ones(1)
tangent = torch.ones(1)

with fwAD.dual_level():
    y = fwAD.make_dual(WrapperTensor(primal), tangent)
    print(y)
```

This prints out:

```
OpOverloadPacket(op='aten.alias')
OpOverloadPacket(op='aten._has_same_storage_numel')
WrapperTensor(self=tensor([1.], tangent=tensor([1.])), elem=tensor([1.]))
```

We see that the WrapperTensor is indeed at the top level, rather than on the inside of the dual number (which is what I originally intended to do with this example).

I suppose this is reasonable semantics. However, the only time I ever wrap tensors is inside `__torch_dispatch__`, so naively I would have expected `make_dual` to show up in the ops that I invoke `__torch_dispatch__` with. But no such op shows up. What it looks like is happening is alias is called on the primal, and the dual number is mashed into this new tensor via some internal API. This means it's not possible to get an accurate trace from the logic here.

**Comment by jbschlosser** [comment]
> This links to Joel's discussion about printing though: how can we provide a good default printing so that users do not hide this information mistakenly. (@jbschlosser did you open an issue about this?)

Opened an issue in #73665; feel free to add discussion there :)

**Comment by soulitzer** [comment]
> The idea here is that you are tracing "below autograd" and so this is expected that you don't see the make_dual.

make_dual is not composite though (anymore)

We could make make_dual composite again though if we also introduce a detach that is backward differentiable but not forward differentiable.

Perhaps we should do that?

**Comment by albanD** [comment]
> make_dual is not composite though

That doesn't change anything :)
